crates/db_schema/src/impls/user_captcha.rs

```diff
@@ -0,0 +1,119 @@
+  schema::user_captcha,
+  source::user_captcha::{CheckCaptcha, UserCaptcha},
+  utils::{functions::lower, get_conn, naive_now, DbPool},
+};
+use diesel::{delete, insert_into, result::Error, ExpressionMethods, QueryDsl};
+use diesel_async::RunQueryDsl;
+
+impl UserCaptcha {
+  pub async fn insert(pool: &DbPool, captcha: &UserCaptcha) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+
+    insert_into(user_captcha::table)
+      .values(captcha)
+      .get_result::<Self>(conn)
+      .await
+  }
+
+  pub async fn check_captcha(pool: &DbPool, to_check: CheckCaptcha) -> Result<bool, Error> {
+    let conn = &mut get_conn(pool).await?;
+
+    // delete any expired captchas
+    delete(user_captcha::table.filter(user_captcha::expires.lt(&naive_now())))
+      .execute(conn)
+      .await?;
+
+    // fetch requested captcha
+    let captcha = user_captcha::dsl::user_captcha
+      .filter((user_captcha::dsl::uuid).eq(to_check.uuid.clone()))
+      .filter(lower(user_captcha::dsl::answer).eq(to_check.answer.to_lowercase().clone()))
+      .first::<Self>(conn)
+      .await;
+
+    let result = captcha.is_ok();
+
+    // delete checked captcha
+    delete(user_captcha::table.filter(user_captcha::uuid.eq(to_check.uuid.clone())))
+      .execute(conn)
+      .await?;
+
+    Ok(result)
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use crate::{
+    source::user_captcha::{CheckCaptcha, UserCaptcha},
+    utils::{build_db_pool_for_tests, naive_now},
+  };
+  use chrono::Duration;
+  use serial_test::serial;
+
+  #[tokio::test]
+  #[serial]
+  async fn test_captcha() {
+    let pool = &build_db_pool_for_tests().await;
+
+    let captcha_a_id = "a".to_string();
+
+    let _ = UserCaptcha::insert(
+      pool,
+      &UserCaptcha {
+        uuid: captcha_a_id.clone(),
+        answer: "XYZ".to_string(),
+        expires: naive_now() + Duration::minutes(10),
+      },
+    )
+    .await;
+
+    let result = UserCaptcha::check_captcha(
+      pool,
+      CheckCaptcha {
+        uuid: captcha_a_id.clone(),
+        answer: "xyz".to_string(),
+      },
+    )
+    .await;
+
+    assert!(result.is_ok());
+    assert!(result.unwrap());
+
+    let result_repeat = UserCaptcha::check_captcha(
+      pool,
+      CheckCaptcha {
+        uuid: captcha_a_id.clone(),
+        answer: "xyz".to_string(),
+      },
+    )
+    .await;
+
+    assert!(result_repeat.is_ok());
+    assert!(!result_repeat.unwrap());
+
+    let expired_id = "already_expired".to_string();
+
+    let _ = UserCaptcha::insert(
+      pool,
+      &UserCaptcha {
+        uuid: expired_id.clone(),
+        answer: "xyz".to_string(),
+        expires: naive_now() - Duration::seconds(1),
+      },
+    )
+    .await;
+
+    let expired_result = UserCaptcha::check_captcha(
+      pool,
+      CheckCaptcha {
+        uuid: expired_id.clone(),
+        answer: "xyz".to_string(),
+      },
+    )
+    .await;
+
+    assert!(expired_result.is_ok());
+    assert!(!expired_result.unwrap());
+  }
```

Comment:
Good tests, but please split them into separate functions so its easy to tell which one is failing.